astropy/cosmology/flrw/base.py

```diff
@@ -194,14 +194,48 @@ def critical_density(self, z: Quantity | ArrayLike) -> Quantity:
         return self.critical_density0 * self.efunc(z) ** 2
 
 
+class _BaryonComponent:
+    """The cosmology has attributes and methods for the baryon density."""
+
+    Ob0: float
+    """Omega baryons: density of baryonic matter in units of the critical density at z=0."""
+
+    inv_efunc: Callable[[NDArray[Any]], NDArray[Any]]
+
+    @deprecated_keywords("z", since="7.0")
+    def Ob(self, z: Quantity | ArrayLike) -> NDArray[Any] | float:
```

Comment:
How simple would it be to replace the type union in the return annotation with overloads?

Reply:
Wouldn't that mean breaking apart ArrayLike?
My plan is to actually remove the `float` option and always return an Array. But This PR shouldn't change API.

Reply:
If you are planning to change the API anyways then introducing the overloads is not worth the effort.